### ISSUE TITLE
Allow colons in bucket and key names

### ISSUE THREAD

**Issue by mtiller** [nats-io/nats.go#1157]
## Feature Request

Allow the use of `:` in bucket names

#### Use Case:

We have a plan to deploy NATS in to support our existing services.  Our hope is to have a single NATS
server that supports various services running in the same data center.  As such, we need to be careful
about naming things.  We had in mind to use URNs to name things like streams and buckets.  The idea
behind using URNs is just to use some existing, well established scheme.

The naming guidelines for streams exclude `.`, `>` and `*`.  So it appears that using these names for 
streams should be fine.  And, indeed, creating a stream with `:` doesn't seem to present any problems.  The
same appears to be true for subjects as well.  But the store is different for buckets.  But I can't see why
since bucket names are just used to construct stream names from what I can tell and they don't seem 
to depend on the use of `:` characters either, _i.e.,_ the `:` doesn't appear in any of these format strings:

```
const (
	kvBucketNamePre         = "KV_"
	kvBucketNameTmpl        = "KV_%s"
	kvSubjectsTmpl          = "$KV.%s.>"
	kvSubjectsPreTmpl       = "$KV.%s."
	kvSubjectsPreDomainTmpl = "%s.$KV.%s."
	kvNoPending             = "0"
)
```

#### Proposed Change:

A simple fix would be to modify the regular expressions associated with bucket and key names by
simply adding the `:` character to the set of permitted characters, _i.e.,_

```
var (
	validBucketRe = regexp.MustCompile(`\A[a-zA-Z0-9_-:]+\z`)
	validKeyRe    = regexp.MustCompile(`\A[:-/_=\.a-zA-Z0-9]+\z`)
)
```

#### Who Benefits From The Change(s)?

Anybody wanting to use URNs as names for buckets or keys.  But more broadly, I'm sure there may be other naming schemes that might also employ a `:` character as well.

The bottom line for me is that I don't see any practical reason to *exclude* `:`s and by allowing them you just have more expressiveness in naming things which seems like a good thing.

#### Alternative Approaches

I recognize that I could just replace `:` with `_` in all cases.  But then if I want to parse these things as URNs I have to transform them back and forth and then all `_` would be treated as `:` (i.e., they couldn't practically exist as two separate characters, _i.e.,_ a URN of the form `urn:bucket:test_cases` really can't be used because round-tripping it like this will result in `urn:bucket:test:cases`).

**Comment by caleblloyd** [comment]
It would make sense that Bucket Name Validation matches Stream Name Validation though, since the Bucket Name ends up being `KV_<bucket_name>`.  Nothing prevents a user from creating a Stream named `KV_a:b:c` using `nats stream create` then having it show up in the `nats kv list`

For FS restrictions, such as those imposed by NTFS on windows, I assume that piece that should be a server and not a client limitation?

**Comment by ripienaar** [comment]
I believe the server now prevents : in stream names. 

Regardless KV spec is deliberately limited in what’s valid so everyone knows what they get and everyone knows what is supported with no weird side effects. Also we plan to eventually support other backends and we surveyed the landscape a bit for a common allowed set. 

These showing up in KV list is a bug imo - KV clients never had a list so I made one for the CLI and it’s too permissive. KV clients now has list and this might behave better.

**Comment by ripienaar** [comment]
You would not be able to access data in buckets made using stream add with a : in them. 

Strongly would not recommend making KV buckets with anything but KV APIs. I have been burned a few times myself by doing that

**Comment by caleblloyd** [comment]
True.

The KV bucket and key constraints are documented in the [ADR-8 spec](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-8.md) so this is not really an issue that could be changed in the Go client without a modification to the spec.

So we should close this issue, and further discussion should there need to be any should be opened against the https://github.com/nats-io/nats-architecture-and-design repo with a reference to ADR-8

**Comment by mtiller** [comment]
My solution here is to use something like base64/62/32 encoded URNs as bucket and key names.  I assume you see no issues with that?

**Comment by ripienaar** [comment]
No, the allowed character list specifically includes all of base64 charset for this reason.
